### PR TITLE
[AMDGPU] Fix kernarg preloading crash with some types and alignments

### PR DESCRIPTION
**Pull request by kerbowa**
Lowering of preloded arguments would fail with half/bfloat if they were dword aligned in the kernarg segment and not part of a vector. Added more tests with different alignments and types.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/preload-kernargs.ll [hunk 1]
```diff
@@ -2317,3 +1408,804 @@ define amdgpu_kernel void @f64_kernel_preload_arg(ptr addrspace(1) %out, double
...
+; GFX90a-PRELOAD-8-NEXT:  ; %bb.0:
+; GFX90a-PRELOAD-8-NEXT:    v_mov_b32_e32 v0, 0
+; GFX90a-PRELOAD-8-NEXT:    v_mov_b32_e32 v1, s8
+; GFX90a-PRELOAD-8-NEXT:    global_store_short v0, v1, s[6:7]
+; GFX90a-PRELOAD-8-NEXT:    s_endpgm
+  store half %in, ptr addrspace(1) %out
+  ret void
+}
+
+define amdgpu_kernel void @bfloat_kernel_preload_arg(ptr addrspace(1) %out, bfloat %in) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/preload-kernargs.ll:1472]: Also test the vectors of half/bfloat/i16?

**Review by arsenm — COMMENTED** [review]
GlobalISel path probably needs equivalent fix? Was this implemented there?

**Comment by kerbowa** [comment]
> GlobalISel path probably needs equivalent fix? Was this implemented there?
Preloading isn't implemented in GlobalISel so the fix isn't needed there. It doesn't crash or anything if you try to use the preloading flag with gisel, but it wont actually do anything.

File: llvm/test/CodeGen/AMDGPU/preload-kernargs.ll (hunk 1 above)
Comment (kerbowa) [reply llvm/test/CodeGen/AMDGPU/preload-kernargs.ll:1472]: There were already tests for vectors of half/i16, it was only raw half type when dword aligned that was crashing.

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/preload-kernargs.ll:1472]: dword aligned bfloat or i16 is tested?

**Review by arsenm — APPROVED** [review]

**Pull request merged**